specification/scripts/simplify_specification.dart

```diff
@@ -0,0 +1,442 @@
+library;
+
+import 'dart:io';
+
+void main() {
+  final inputFile = File(specificationFilename);
+  if (!inputFile.existsSync()) fail("Specification not found");
+  final contents = inputFile.readAsLinesSync();
+  final workingContents =
+      List<String?>.from(contents, growable: false)
+        ..removeComments()
+        ..removeDartCode()
+        ..removeTrailingWhitespace()
+        ..removeNonNormative()
+        ..joinLines()
+        ..removeTrailingWhitespace()
+        ..removeSpuriousEmptyLines();
+  final simplifiedContents = workingContents.whereType<String>().toList();
+  final outputFile = File(outputFilename);
+  final outputSink = outputFile.openWrite();
+  simplifiedContents.forEach(outputSink.writeln);
+}
+
+const outputFilename = 'dartLangSpec-simple.tex';
+
+const specificationFilename = 'dartLangSpec.tex';
+
+void fail(String message) {
+  print("simplify_specification error: $message");
+  exit(-1);
+}
+
+extension on String {
+  bool get endsCode => startsWith(r"\end{normativeDartCode}");
+  bool get endsList =>
+      startsWith(r"\end{itemize}") || startsWith(r"\end{enumerate}");
+  bool get endsMinipage => contains(r"\end{minipage}");
+  bool get isItem => startsWith(r"\item");
+  bool get startsCode => startsWith(r"\begin{normativeDartCode}");
+  bool get startsList =>
+      startsWith(r"\begin{itemize}") || startsWith(r"\begin{enumerate}");
+  bool get startsMinipage => contains(r"\begin{minipage}");
+}
+
+extension on List<String?> {
+  static final _commentRegExp = RegExp(r"^%|[^%\\]%");
+  static final _commentaryRationaleRegExp = RegExp(
+    r"^ *\(?\\(commentary|rationale){",
+  );
+  static final _bracesRegExp = RegExp(r"\\[a-zA-Z]*{.*}");
+  static final _parenBracesRexExp = RegExp(r"\(\\[a-zA-Z]*{.*}\)");
+
+  void joinLines() {
+    bool inFrontMatter = true;
+    for (var lineIndex = 0; lineIndex < length; ++lineIndex) {
+      final line = this[lineIndex]; // Invariant.
+      if (line == null) continue;
+      if (inFrontMatter) {
+        if (line.startsWith(r"\begin{document}")) inFrontMatter = false;
+        continue;
+      }
+      if (line.startsWith(r"\LMHash{}")) {
+        lineIndex = _gatherParagraph(lineIndex);
+      } else if (line.startsWith(r"\noindent")) {
+        lineIndex = _gatherContinuedParagraph(lineIndex);
+      } else if (line.startsList) {
+        lineIndex = _gatherItems(lineIndex);
+      }
+    }
+  }
+
+  // Eliminate comment-only lines. Reduce other comments to `%`.
+  void removeComments() {
+    for (int i = 0; i < length; ++i) {
+      final line = this[i];
+      if (line == null) continue; // It isn't, but flow-analysis doesn't know.
+      final match = _commentRegExp.firstMatch(line);
+      if (match != null) {
+        if (match.start == 0) {
+          this[i] = null; // A comment-only line disappears entirely.
+        } else {
+          final cutPosition = match.start + 2; // Include the `%`.
+          if (line.trimLeft().codeUnitAt(0) == 37) {
```

Comment:
Nit: Is there a more descriptive way to describe the code unit of '%'?

Reply:
Introduced a constant variable with a descriptive name.